src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqTouchstoneRepository.kt

```diff
@@ -191,14 +191,75 @@ class JooqTouchstoneRepository(
         return records.map { mapCoverageSet(it) }
     }
 
+    private fun coverageDimensions() : Array<Field<*>>
+    {
+        //The columns in the Coverage table which the target and coverage values are grouped by
+        return arrayOf(COVERAGE.COVERAGE_SET,
+                COVERAGE.YEAR,
+                COVERAGE.COUNTRY,
+                COVERAGE.AGE_FROM,
+                COVERAGE.AGE_TO,
+                COVERAGE.AGE_RANGE_VERBATIM,
+                COVERAGE.GAVI_SUPPORT,
+                COVERAGE.GENDER)
+    }
+
+    private fun validTargetOrNull() : Field<BigDecimal?>
+    {
+        //Coverage and target values only make sense if both are provided. Return the target value of a row providing
+        //coverage is not null, else return null
+        return `when`(COVERAGE.COVERAGE_.isNull(), COVERAGE.COVERAGE_) //This is actually the easiest way to get a known NULL here!
+                .otherwise(COVERAGE.TARGET)
+    }
+
+    private fun validCoverageOrNull() : Field<BigDecimal?>
+    {
+        //Coverage and target values only make sense if both are provided. Return the coverage value of a row providing
+        //target is not null, else return null
+        return `when`(COVERAGE.TARGET.isNull(), COVERAGE.TARGET) //This is actually the easiest way to get a known NULL here!
+                .otherwise(COVERAGE.COVERAGE_)
+    }
+
+    private fun aggregatedValues() : List<Field<*>>
+    {
+        return arrayOf(
+                //Aggregated coverage - the sum of each row's coverage * target (to get total no of fvp's across campaign)
+                //divided by the the total target population
+
```

Comment:
It occurred to me that we could maybe use a user-defined function to calculate these rounded values - in a function you could store the aggregatedCoverage result in a declared variable so you wouldn't need to keep recalculating it in a single SQL expression. Not sure if we're already using UDFs elsewhere? Of course they have a maintenance overhead, and potentially a performance overhead of their own. I assume you can use a function on all the rows of a result set? You could in SQL Server. Maybe one to try if performance of this solution is poor...